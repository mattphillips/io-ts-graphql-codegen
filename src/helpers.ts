export const renderHelpers = `
import * as t from 'io-ts';
import { optionFromNullable } from 'io-ts-types';

const Pick = <P extends t.Props, Key extends keyof P>(p: P, ...keys: Array<Key>): t.TypeC<Pick<P, Key>> => {
  const entries = Object.keys(p).map(k => [k, p[k]]) as Array<[Key, t.mixed]>;
  const props = entries.reduce<Pick<P, Key>>(
    (acc, [key, value]) => (keys.includes(key) ? { ...acc, [key]: value } : acc),
    {} as Pick<P, Key>
  );

  return t.type(props);
};
`;
