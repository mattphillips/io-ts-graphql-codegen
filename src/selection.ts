import { OperationDefinitionNode, SelectionNode, NameNode, FieldNode, InlineFragmentNode } from 'graphql';
import {
  ObjectType,
  IoTsType,
  IntersectionType,
  Typename,
  PickType,
  isNamedType,
  ArrayType,
  OptionType,
  UnionType,
  MutationType,
  QueryType,
} from './IoTsType';

export const make = (operation: OperationDefinitionNode, root: QueryType | MutationType): QueryType | MutationType => {
  const go = (node: SelectionNode, lastVisited: IoTsType): IoTsType => {
    if (node.kind === 'Field') {
      const name = getName(node);
      const current = getProps(lastVisited)[name];

      if (current === undefined) {
        throw new Error(`Field: ${name} does not exist on type: ${lastVisited.tag}`);
      }

      const primitiveSelections =
        node.selectionSet?.selections.filter(
          (s): s is FieldNode => s.kind === 'Field' && s.selectionSet === undefined && getName(s) !== '__typename'
        ) || [];

      const nestedSelections =
        node.selectionSet?.selections.filter(
          (s): s is FieldNode => s.kind === 'Field' && s.selectionSet !== undefined
        ) || [];

      const unionSelections =
        node.selectionSet?.selections.filter(
          (s): s is InlineFragmentNode => s.kind === 'InlineFragment' && s.selectionSet !== undefined
        ) || [];

      if (current.tag === 'Named' && current.value.tag === 'Union' && unionSelections.length > 0) {
        const missingCases = current.value.types
          .filter(isNamedType)
          .map((type) => type.name)
          .filter(
            (name) => unionSelections.find((selection) => selection.typeCondition?.name.value === name) === undefined
          );

        if (missingCases.length > 0) {
          throw new Error(
            `Non exhaustive matching on Union: \`${current.name}\`. \n\nMissing cases: \n\t- ${missingCases.join(
              '\n\t- '
            )}`
          );
        }

        return UnionType(unionSelections.map((s) => go(s, current)));
      }

      if (current.tag === 'Scalar') {
        return current;
      }

      if (current.tag === 'Array') {
        return ArrayType(
          IntersectionType([
            Typename(getIoTsName(current.value)),
            PickType(getIoTsName(current.value), ...primitiveSelections.map((s) => getName(s))),
            ...nestedSelections.map((s) => ObjectType({ [getName(s)]: go(s, current.value) })),
          ])
        );
      }

      if (current.tag === 'Option') {
        return OptionType(
          IntersectionType([
            Typename(getIoTsName(current.value)),
            PickType(getIoTsName(current.value), ...primitiveSelections.map((s) => getName(s))),
            ...nestedSelections.map((s) => ObjectType({ [getName(s)]: go(s, current.value) })),
          ])
        );
      }

      if (nestedSelections.length === 0 && primitiveSelections.length === 0) {
        return IntersectionType([Typename(getIoTsName(lastVisited)), PickType(getIoTsName(lastVisited), name)]);
      }

      return IntersectionType([
        Typename(getIoTsName(current)),
        PickType(getIoTsName(current), ...primitiveSelections.map((s) => getName(s))),
        ...nestedSelections.map((s) => ObjectType({ [getName(s)]: go(s, current) })),
      ]);
    }

    if (node.kind === 'InlineFragment') {
      const current =
        lastVisited.tag === 'Named' && lastVisited.value.tag === 'Union'
          ? lastVisited.value.types.find((s) => s.tag === 'Named' && s.name === getName(node.typeCondition))
          : undefined;

      if (current === undefined) {
        throw new Error(`Could not find ${getName(node.typeCondition)} in union type: ${getIoTsName(lastVisited)}`);
      }

      const primitiveSelections =
        node.selectionSet?.selections.filter(
          (s): s is FieldNode => s.kind === 'Field' && s.selectionSet === undefined && getName(s) !== '__typename'
        ) || [];

      const nestedSelections =
        node.selectionSet?.selections.filter(
          (s): s is FieldNode => s.kind === 'Field' && s.selectionSet !== undefined
        ) || [];

      return IntersectionType([
        Typename(getIoTsName(current)),
        PickType(getIoTsName(current), ...primitiveSelections.map((s) => getName(s))),
        ...nestedSelections.map((s) => ObjectType({ [getName(s)]: go(s, current) })),
      ]);
    }

    throw 'not supported yet';
  };

  const TypeConstructor = operation.operation === 'mutation' ? MutationType : QueryType;

  return TypeConstructor(
    getName(operation),
    operation.selectionSet.selections
      .filter(isField)
      .map<[string, IoTsType]>((s) => [getName(s), go(s, root)])
      .reduce((acc, [key, value]) => ({ ...acc, [key]: value }), {})
  );
};

const getProps = (node: IoTsType) => {
  if (node.tag === 'Mutation' || node.tag === 'Query' || node.tag === 'Object') return node.props;
  if (node.tag === 'Named' && node.value.tag === 'Object') return node.value.props;
  return {};
};

const getIoTsName = (type: IoTsType) => (isNamedType(type) ? type.name : '');
const getName = (node: { name: NameNode } | { name?: { value: string } } | undefined): string =>
  node?.name?.value || '';
const isField = (s: SelectionNode): s is FieldNode => s.kind === 'Field';
