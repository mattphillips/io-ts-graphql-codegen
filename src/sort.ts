import {
  ObjectTypeDefinitionNode,
  UnionTypeDefinitionNode,
  NamedTypeNode,
  EnumTypeDefinitionNode,
  InputObjectTypeDefinitionNode,
  TypeNode,
  DefinitionNode,
  ScalarTypeDefinitionNode
} from 'graphql';

export type TypeDefinition =
  | ObjectTypeDefinitionNode
  | UnionTypeDefinitionNode
  | EnumTypeDefinitionNode
  | InputObjectTypeDefinitionNode
  | ScalarTypeDefinitionNode;

type Graph = ReadonlyArray<TypeDefinition>;
type InGraph = ReadonlyArray<TypeDefinition | NamedTypeNode>;
type Visited = Set<TypeDefinition>;
type State = [Visited, Graph];

export const sortGraphTypes = (graph: InGraph): Graph => {
  const goAll = (g: InGraph, initVisited: Visited, initAcc: Graph) =>
    g.reduce<State>(([visited, acc], node) => go(node, visited, acc), [initVisited, initAcc]);

  const go = (node: TypeDefinition | NamedTypeNode, visited: Visited, acc: Graph): State => {
    if ([...visited.values()].some(v => v.name.value === node.name.value)) {
      return [visited, acc];
    }

    if (node.kind === 'ObjectTypeDefinition' && node.fields!.length > 0) {
      const [updatedVisted, updatedAcc] = goAll(
        node.fields?.map(f => getType(f.type)) || [],
        new Set([...visited, node]),
        acc
      );
      return [updatedVisted, updatedAcc.concat(node)];
    }

    if (node.kind === 'UnionTypeDefinition' && node.types!.length > 0) {
      const [updatedVisted, updatedAcc] = goAll(node.types!, new Set([...visited, node]), acc);
      return [updatedVisted, updatedAcc.concat(node)];
    }

    if (node.kind === 'EnumTypeDefinition') {
      return [new Set([...visited.values(), node]), acc.concat(node)];
    }

    if (node.kind === 'InputObjectTypeDefinition') {
      const [updatedVisted, updatedAcc] = goAll(
        node.fields?.map(f => getType(f.type)) || [],
        new Set([...visited, node]),
        acc
      );
      return [updatedVisted, updatedAcc.concat(node)];
    }

    if (node.kind === 'ScalarTypeDefinition') {
      return [new Set([...visited.values(), node]), acc.concat(node)];
    }

    // Check if the current NamedType is an ObjectTypeDefinition or UnionTypeDefinition
    const maybeNode = graph.find(v => v.name.value === node.name.value);

    if (maybeNode !== undefined) {
      return go(maybeNode, visited, acc);
    }

    if (node.kind === 'NamedType') {
      return [visited, acc];
    }

    return [new Set([...visited.values(), node]), acc.concat(node)];
  };

  return goAll(graph, new Set(), [])[1];
};

const getType = (node: TypeNode): NamedTypeNode => {
  switch (node.kind) {
    case 'ListType': {
      return getType(node.type);
    }

    case 'NamedType': {
      return node;
    }

    case 'NonNullType': {
      return getType(node.type);
    }
  }
};

export const isTypeDefinition = (
  node: DefinitionNode
): node is
  | ObjectTypeDefinitionNode
  | UnionTypeDefinitionNode
  | EnumTypeDefinitionNode
  | InputObjectTypeDefinitionNode =>
  node.kind === 'ObjectTypeDefinition' ||
  node.kind === 'UnionTypeDefinition' ||
  node.kind === 'EnumTypeDefinition' ||
  node.kind === 'InputObjectTypeDefinition' ||
  node.kind === 'ScalarTypeDefinition';
