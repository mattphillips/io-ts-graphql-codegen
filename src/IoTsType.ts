import { TypeNode } from 'graphql';
import { TypeDefinition } from './sort';

export type IoTsType =
  | IntersectionType
  | UnionType
  | ObjectType
  | LiteralType
  | PickType
  | ArrayType
  | EnumType
  | OptionType
  | ScalarType
  | PrimitiveType
  | InputObjectType
  | QueryType
  | MutationType
  | Named<UnionType>
  | Named<ObjectType>;

export type IntersectionType = { tag: 'Intersection'; types: Array<IoTsType> };
export const IntersectionType = (types: Array<IoTsType>): IntersectionType => ({ tag: 'Intersection', types });

export type UnionType = { tag: 'Union'; types: Array<IoTsType> };
export const UnionType = (types: Array<IoTsType>): UnionType => ({
  tag: 'Union',
  types,
});

export type ObjectType = { tag: 'Object'; props: Record<string, IoTsType> };
export const ObjectType = (props: Record<string, IoTsType>): ObjectType => ({
  tag: 'Object',
  props,
});

export type Named<A extends UnionType | ObjectType> = { tag: 'Named'; value: A; name: string };
export const Named = <A extends UnionType | ObjectType>(name: string, value: A): Named<A> => ({
  tag: 'Named',
  name,
  value,
});

export type InputObjectType = { tag: 'Input'; props: Record<string, IoTsType>; name: string };
export const InputObjectType = (name: string, props: Record<string, IoTsType>): InputObjectType => ({
  tag: 'Input',
  props,
  name,
});

export type QueryType = { tag: 'Query'; props: Record<string, IoTsType>; name: string };
export const QueryType = (name: string, props: Record<string, IoTsType>): QueryType => ({
  tag: 'Query',
  props,
  name,
});

export type MutationType = { tag: 'Mutation'; props: Record<string, IoTsType>; name: string };
export const MutationType = (name: string, props: Record<string, IoTsType>): MutationType => ({
  tag: 'Mutation',
  props,
  name,
});

export type LiteralType = { tag: 'Literal'; value: string };
export const LiteralType = (value: string): LiteralType => ({ tag: 'Literal', value });

export type ArrayType = { tag: 'Array'; value: IoTsType };
export const ArrayType = (value: IoTsType): ArrayType => ({ tag: 'Array', value });

export type ScalarType = { tag: 'Scalar'; name: string };
export const ScalarType = (name: string): ScalarType => ({ tag: 'Scalar', name });

export type StringType = { tag: 'String' };
export const StringType: StringType = { tag: 'String' };

export type IntType = { tag: 'Int' };
export const IntType: IntType = { tag: 'Int' };

export type BooleanType = { tag: 'Boolean' };
export const BooleanType: BooleanType = { tag: 'Boolean' };

export type FloatType = { tag: 'Float' };
export const FloatType: FloatType = { tag: 'Float' };

export type IdType = { tag: 'ID' };
export const IdType: IdType = { tag: 'ID' };

export type PrimitiveType = StringType | IntType | BooleanType | FloatType | IdType;

export const isPrimitive = (ast: IoTsType): ast is PrimitiveType =>
  ast.tag === 'String' || ast.tag === 'Boolean' || ast.tag === 'Int' || ast.tag === 'ID' || ast.tag === 'Float';

export type EnumType = { tag: 'Enum'; cases: Array<string>; name: string };
export const EnumType = (name: string, ...cases: Array<string>): EnumType => ({
  tag: 'Enum',
  cases,
  name,
});

export type OptionType = { tag: 'Option'; value: IoTsType };
export const OptionType = (value: IoTsType): OptionType => ({ tag: 'Option', value });

export type PickType = { tag: 'Pick'; type: string; fields: Array<string> };
export const PickType = (type: string, ...fields: Array<string>): PickType => ({ tag: 'Pick', type, fields });

export type NamedType =
  | Named<UnionType>
  | Named<ObjectType>
  | EnumType
  | ScalarType
  | InputObjectType
  | QueryType
  | MutationType;

export const Typename = (name: string): Named<ObjectType> => Named(name, ObjectType({}));

export const isNamedType = (type: IoTsType): type is NamedType =>
  type.tag === 'Named' ||
  type.tag === 'Object' ||
  type.tag === 'Scalar' ||
  type.tag === 'Enum' ||
  type.tag === 'Input' ||
  type.tag === 'Query' ||
  type.tag === 'Mutation';

export const DEFAULT_SCALARS = new Map<string, PrimitiveType>([
  ['ID', IdType],
  ['String', StringType],
  ['Boolean', BooleanType],
  ['Int', IntType],
  ['Float', FloatType],
]);
export const make = (node: TypeDefinition, namedTypes: Map<string, TypeDefinition>): IoTsType => {
  const name = node.name.value;
  const getScalar = (name: string) => {
    if (DEFAULT_SCALARS.has(name)) {
      return DEFAULT_SCALARS.get(name)!;
    }

    // TODO: this could fail
    return make(namedTypes.get(name)!, namedTypes);
  };

  const getType = (type: TypeNode, required = false): IoTsType => {
    switch (type.kind) {
      case 'ListType': {
        if (required) {
          return ArrayType(getType(type.type));
        }
        return OptionType(ArrayType(getType(type.type)));
      }

      case 'NonNullType': {
        return getType(type.type, true);
      }

      case 'NamedType': {
        const t = getScalar(type.name.value);
        if (required) {
          return t;
        }
        return OptionType(t);
      }
    }
  };

  switch (node.kind) {
    case 'EnumTypeDefinition': {
      return EnumType(node.name.value, ...(node.values?.map((v) => v.name.value) || []));
    }

    case 'ObjectTypeDefinition': {
      const props =
        node.fields?.reduce(
          (acc, field) => ({ ...acc, [field.name.value]: getType(field.type) }),
          {} as Record<string, IoTsType>
        ) || {};

      if (name === 'Query') return QueryType(name, props);
      if (name === 'Mutation') return MutationType(name, props);
      return Named(name, ObjectType(props));
    }

    case 'InputObjectTypeDefinition': {
      return InputObjectType(
        name,
        node.fields?.reduce(
          (acc, field) => ({ ...acc, [field.name.value]: getType(field.type) }),
          {} as Record<string, IoTsType>
        ) || {}
      );
    }

    case 'ScalarTypeDefinition': {
      return ScalarType(name);
    }

    case 'UnionTypeDefinition': {
      return Named(
        name,
        UnionType(
          node.types?.map((t) => {
            const type = namedTypes.get(t.name.value);
            if (type === undefined) {
              throw new Error(`Could not find type definition with name: ${t.name.value}`);
            }
            return make(type, namedTypes);
          }) || []
        )
      );
    }
  }
};
