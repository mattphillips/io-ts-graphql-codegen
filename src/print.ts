import { IoTsType, NamedType, isNamedType, InputObjectType, MutationType, QueryType, OptionType } from './IoTsType';

const chunkArray = <A>(n: number, as: Array<A>): Array<Array<A>> =>
  as.reduce((acc, a) => {
    const tail = acc[acc.length - 1];

    if (tail && tail.length < n) {
      return [...acc.slice(0, acc.length - 1), [...tail, a]];
    }
    return [...acc, [a]];
  }, [] as Array<Array<A>>);

const mkPrinter = (optionPrinter: (ast: OptionType) => string) => (ast: IoTsType): string => {
  const print = mkPrinter(optionPrinter);

  switch (ast.tag) {
    case 'Array': {
      if (isNamedType(ast.value)) return `t.array(${ast.value.name})`;
      return `t.array(${print(ast.value)})`;
    }

    // NOTE: Currently this doesn't support > 25 types
    case 'Intersection': {
      if (ast.types.length > 5) {
        const chunks = ast.types.length % 5 === 1 ? 4 : 5;
        return `t.intersection([${chunkArray(chunks, ast.types)
          .map(types => `t.intersection([${types.map(print).join(', ')}])`)
          .join(', ')}])`;
      }
      return `t.intersection([${ast.types.map(print).join(', ')}])`;
    }
    case 'Literal': {
      return `t.literal('${ast.value}')`;
    }
    case 'Object':
    case 'Query':
    case 'Mutation':
    case 'Input': {
      return `t.type({ ${Object.entries(ast.props)
        .map(([key, value]) => {
          if (isNamedType(value) && value.name !== '') return `${key}: ${value.name}`;
          return `${key}: ${print(value)}`;
        })
        .join(',\n')}})`;
    }
    case 'Pick': {
      return `Pick(${ast.type}.props, ${ast.fields.map(f => `'${f}'`).join(',')})`;
    }
    case 'Union': {
      return `t.union([${ast.types
        .map(n => {
          if (isNamedType(n)) return n.name;
          return print(n);
        })
        .join(', ')}])`;
    }
    case 'Scalar': {
      return 't.any';
    }
    case 'Option': {
      return optionPrinter(ast);
    }
    case 'Enum': {
      if (ast.cases.length === 1) return `t.literal('${ast.cases[0]}')`;
      return `t.union([${ast.cases.map(c => `t.literal('${c}')`).join(',')}])`;
    }
    case 'String': {
      return 't.string';
    }
    case 'Int': {
      // TODO: should this be t.Int?
      return 't.number';
    }
    case 'Float': {
      return 't.number';
    }
    case 'Boolean': {
      return 't.boolean';
    }
    case 'ID': {
      return 't.string';
    }
    case 'Named': {
      if (ast.value.tag === 'Object') {
        return `t.type({ 
          __typename: t.literal('${ast.name}'),
          ${Object.entries(ast.value.props)
            .map(([key, value]) => {
              if (isNamedType(value) && value.name !== '') return `${key}: ${value.name}`;
              return `${key}: ${print(value)}`;
            })
            .join(',\n')}
        })`;
      }
      return print(ast.value);
    }
  }
};

export const print = mkPrinter((ast: OptionType): string => {
  if (isNamedType(ast.value)) return `optionFromNullable(${ast.value.name})`;
  return `optionFromNullable(${print(ast.value)})`;
});

export const printNamedType = (ast: NamedType): string => {
  if (ast.tag === 'Input') return printVariables(ast);
  return `
    export const ${ast.name} = ${print(ast)};
    export type ${ast.name} = t.TypeOf<typeof ${ast.name}>;
  `;
};

export const printSelection = (ast: MutationType | QueryType): string =>
  `
    export const ${ast.name}${ast.tag} = t.type({ data: ${print(ast)} });
    export type ${ast.name}${ast.tag} = t.TypeOf<typeof ${ast.name}${ast.tag}>['data'];
  `;

export const printVariables = (ast: InputObjectType): string => {
  const optionPrinter = (ast: OptionType): string => {
    if (isNamedType(ast.value)) return `t.union([t.null, ${ast.value.name}])`;
    return `t.union([t.null, ${print(ast.value)}])`;
  };

  return `
    export const ${ast.name} = ${mkPrinter(optionPrinter)(ast)};
    export type ${ast.name} = t.TypeOf<typeof ${ast.name}>;
  `;
};
