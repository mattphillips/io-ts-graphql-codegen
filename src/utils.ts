import { NameNode, DefinitionNode, OperationDefinitionNode, TypeNode } from 'graphql';
import { IoTsType, QueryType, MutationType } from './IoTsType';

export const getName = ({ name }: { name: NameNode }): string => name.value;

export const isOperationDefinitionNode = (node: DefinitionNode): node is OperationDefinitionNode =>
  node.kind === 'OperationDefinition';

export const capitalizeFirstLetter = (s: string): string => s.charAt(0).toUpperCase() + s.slice(1);

const getQuery = (graph: Array<IoTsType>): QueryType => graph.find((node): node is QueryType => node.tag === 'Query')!;

const getMutation = (graph: Array<IoTsType>): MutationType =>
  graph.find((node): node is MutationType => node.tag === 'Mutation')!;

// NOTE: only supports mutations/queries and there must be present
export const getRoot = (operation: OperationDefinitionNode, graph: Array<IoTsType>): QueryType | MutationType =>
  operation.operation === 'mutation' ? getMutation(graph) : getQuery(graph);

export const getTypeName = (type: TypeNode): string => {
  if (type.kind === 'ListType' || type.kind === 'NonNullType') return getTypeName(type.type);
  return type.name.value;
};
