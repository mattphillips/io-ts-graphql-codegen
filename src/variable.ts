import { OperationDefinitionNode, TypeNode } from 'graphql';
import { InputObjectType, ScalarType, PrimitiveType, EnumType, IoTsType, ArrayType, OptionType } from './IoTsType';
import { capitalizeFirstLetter, getName } from './utils';

export const make = (namedInputs: Map<string, InputObjectType | ScalarType | PrimitiveType | EnumType>) => (
  node: OperationDefinitionNode
): InputObjectType => {
  const getType = (type: TypeNode, required = false): IoTsType => {
    switch (type.kind) {
      case 'ListType': {
        if (required) {
          return ArrayType(getType(type.type));
        }
        return OptionType(ArrayType(getType(type.type)));
      }

      case 'NonNullType': {
        return getType(type.type, true);
      }

      case 'NamedType': {
        // TODO: this could fail
        const t = namedInputs.get(type.name.value)!;
        if (required) {
          return t;
        }
        return OptionType(t);
      }
    }
  };

  return InputObjectType(
    `${node.name?.value}${capitalizeFirstLetter(node.operation)}Variables`,
    node.variableDefinitions?.reduce(
      (acc, def) => ({
        ...acc,
        [getName(def.variable)]: getType(def.type),
      }),
      {}
    ) || {}
  );
};
