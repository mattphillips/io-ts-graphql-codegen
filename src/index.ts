import { Types } from '@graphql-codegen/plugin-helpers';
import { GraphQLSchema, printSchema, parse, DefinitionNode } from 'graphql';
import { format } from 'prettier';

import { renderHelpers } from './helpers';
import { make, isNamedType, InputObjectType, ScalarType, PrimitiveType, DEFAULT_SCALARS, EnumType } from './IoTsType';
import { printNamedType, printSelection, printVariables } from './print';
import { make as makeSelection } from './selection';
import { sortGraphTypes, isTypeDefinition } from './sort';
import { isOperationDefinitionNode, getRoot } from './utils';
import { make as makeVariable } from './variable';

export const plugin = (schema: GraphQLSchema, documents: Types.DocumentFile[]): string => {
  const printedSchema = printSchema(schema);
  const { definitions } = parse(printedSchema);

  const documentDefinitions = documents
    .reduce<ReadonlyArray<DefinitionNode>>((acc, { document }) => [...acc, ...document!.definitions], [])
    .filter(isOperationDefinitionNode);

  const graph = sortGraphTypes(definitions.filter(isTypeDefinition));
  const namedTypes = graph.reduce((acc, type) => acc.set(type.name.value, type), new Map());
  const ioTsAst = graph.map((n) => make(n, namedTypes)).filter(isNamedType);

  const selections = documentDefinitions.map((def) => makeSelection(def, getRoot(def, ioTsAst)));

  const namedInputs = ioTsAst
    .filter(
      (ast): ast is ScalarType | InputObjectType | EnumType =>
        ast.tag === 'Scalar' || ast.tag === 'Input' || ast.tag === 'Enum'
    )
    .reduce(
      (acc, node) => acc.set(node.name, node),
      new Map<string, ScalarType | InputObjectType | PrimitiveType | EnumType>(DEFAULT_SCALARS.entries())
    );

  const variables = documentDefinitions.filter(isOperationDefinitionNode).map(makeVariable(namedInputs));

  return format(
    `
    ${renderHelpers}

    ${ioTsAst.map(printNamedType).join('\n')}

    ${variables.map(printVariables).join('\n')}

    ${selections.map(printSelection).join('\n')}
  `,
    { parser: 'typescript', printWidth: 120 }
  );
};
