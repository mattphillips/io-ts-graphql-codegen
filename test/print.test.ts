import {
  IntersectionType,
  ObjectType,
  Typename,
  UnionType,
  Named,
  QueryType,
  MutationType,
  ArrayType,
  ScalarType,
  PickType,
  StringType,
  IntType,
  LiteralType,
  InputObjectType,
  OptionType,
  FloatType,
  BooleanType,
  IdType
} from 'src/IoTsType';
import { print, printNamedType } from 'src/print';

describe('Print', () => {
  describe('Array', () => {
    it('should return named values inlined with name', () => {
      expect(print(ArrayType(ScalarType('Named')))).toEqual('t.array(Named)');
    });

    it('should return non-named values completely inlined', () => {
      expect(print(ArrayType(PickType('Named', 'a')))).toEqual("t.array(Pick(Named.props, 'a'))");
    });
  });

  describe('Intersection', () => {
    it('should return intersection types inlined', () => {
      expect(print(IntersectionType([StringType, IntType]))).toEqual(`t.intersection([t.string, t.number])`);
    });

    it('should return types inlined when given types list size === 5', () => {
      expect(print(IntersectionType([StringType, IntType, BooleanType, FloatType, IdType]))).toEqual(
        `t.intersection([t.string, t.number, t.boolean, t.number, t.string])`
      );
    });

    it('should return nested intersections when given type list size === 7 with a max size of 5 types per intersection', () => {
      expect(
        print(IntersectionType([StringType, StringType, StringType, StringType, StringType, StringType, StringType]))
      ).toEqual(
        `t.intersection([t.intersection([t.string, t.string, t.string, t.string, t.string]), t.intersection([t.string, t.string])])`
      );
    });
    it('should return nested intersections when given type list size === 6 with a max size of 4 types per intersection', () => {
      expect(print(IntersectionType([StringType, StringType, StringType, StringType, StringType, StringType]))).toEqual(
        `t.intersection([t.intersection([t.string, t.string, t.string, t.string]), t.intersection([t.string, t.string])])`
      );
    });

    it('should return nested intersections when given type list size is greater than 5', () => {
      expect(
        print(
          IntersectionType([
            StringType,
            StringType,
            StringType,
            StringType,
            StringType,
            StringType,
            StringType,
            StringType,
            StringType,
            StringType,
            StringType,
            StringType
          ])
        )
      ).toEqual(
        `t.intersection([t.intersection([t.string, t.string, t.string, t.string, t.string]), t.intersection([t.string, t.string, t.string, t.string, t.string]), t.intersection([t.string, t.string])])`
      );
    });
  });

  describe('Literal', () => {
    it('should render literal with given value', () => {
      expect(print(LiteralType('Value'))).toEqual(`t.literal('Value')`);
    });
  });

  describe.each`
    type          | fn
    ${'Query'}    | ${QueryType}
    ${'Mutation'} | ${MutationType}
    ${'Input'}    | ${InputObjectType}
  `('$type', ({ fn }) => {
    it('should inline non-named types', () => {
      expect(print(fn('', { a: LiteralType('A') }))).toMatchSnapshot();
    });

    it('should inline named types name value', () => {
      expect(print(fn('', { a: ScalarType('A') }))).toMatchSnapshot();
    });
  });

  describe('ObjectType', () => {
    it('should inline non-named types', () => {
      expect(print(ObjectType({ a: LiteralType('A') }))).toMatchSnapshot();
    });

    it('should inline named types name value', () => {
      expect(print(ObjectType({ a: ScalarType('A') }))).toMatchSnapshot();
    });
  });

  describe('Pick', () => {
    it('should return selection inlined', () => {
      expect(print(PickType('Named', 'a'))).toEqual("Pick(Named.props, 'a')");
    });
  });

  describe('Scalar', () => {
    it('should return any codec', () => {
      expect(print(ScalarType('A'))).toEqual('t.any');
    });
  });

  describe('Option', () => {
    it('should return named values inlined with name', () => {
      expect(print(OptionType(ScalarType('Named')))).toEqual('optionFromNullable(Named)');
    });

    it('should return non-named values completely inlined', () => {
      expect(print(OptionType(PickType('Named', 'a')))).toEqual("optionFromNullable(Pick(Named.props, 'a'))");
    });
  });

  describe('String', () => {
    it('should return string codec', () => {
      expect(print(StringType)).toEqual('t.string');
    });
  });

  describe('Float', () => {
    it('should return number codec', () => {
      expect(print(FloatType)).toEqual('t.number');
    });
  });

  describe('Boolean', () => {
    it('should return boolean codec', () => {
      expect(print(BooleanType)).toEqual('t.boolean');
    });
  });

  describe('ID', () => {
    it('should return string codec', () => {
      expect(print(IdType)).toEqual('t.string');
    });
  });

  describe('Int', () => {
    it('should return number codec', () => {
      expect(print(IntType)).toEqual('t.number');
    });
  });

  describe('Union', () => {
    it('should return unions inlined', () => {
      expect(print(UnionType([IntersectionType([Typename('Test')]), IntersectionType([Typename('Test2')])])))
        .toMatchInlineSnapshot(`
        t.union([
          t.intersection([
            t.type({
              __typename: t.literal("Test"),
            }),
          ]),
          t.intersection([
            t.type({
              __typename: t.literal("Test2"),
            }),
          ]),
        ]);

      `);
    });
    it('should return unions named types names inline', () => {
      expect(print(UnionType([ScalarType('A'), ScalarType('B')]))).toMatchInlineSnapshot(`
        t.union([A, B]);

      `);
    });

    it('should return nested unions', () => {
      expect(
        print(
          Named(
            'Blah',
            ObjectType({
              blah: UnionType([IntersectionType([Typename('Test')]), IntersectionType([Typename('Test2')])])
            })
          )
        )
      ).toMatchInlineSnapshot(`
        t.type({
          __typename: t.literal("Blah"),
          blah: t.union([
            t.intersection([
              t.type({
                __typename: t.literal("Test"),
              }),
            ]),
            t.intersection([
              t.type({
                __typename: t.literal("Test2"),
              }),
            ]),
          ]),
        });

      `);
    });
  });

  describe('Named', () => {
    describe('Union', () => {
      it('should return unions without name', () => {
        expect(print(Named('Ignored', UnionType([ScalarType('A'), ScalarType('B')])))).toEqual(`t.union([A, B])`);
      });
    });

    describe('ObjectType', () => {
      it('should inline non-named types with __typename', () => {
        expect(print(Named('Alphabet', ObjectType({ a: LiteralType('A') })))).toMatchSnapshot();
      });

      it('should inline named types name value with __typename', () => {
        expect(print(Named('Alphabet', ObjectType({ a: ScalarType('A') })))).toMatchSnapshot();
      });
    });
  });

  describe('printNamedType', () => {
    describe('InputType', () => {
      it('should remove option types with null union on primitives', () => {
        expect(printNamedType(InputObjectType('Name', { a: OptionType(StringType) }))).toMatchSnapshot();
      });

      it('should remove option types with null union on array primitives', () => {
        expect(printNamedType(InputObjectType('Name', { a: ArrayType(OptionType(StringType)) }))).toMatchSnapshot();
      });

      it('should remove option types with null union on custom inputs', () => {
        expect(
          printNamedType(InputObjectType('Name', { a: OptionType(InputObjectType('A', { b: StringType })) }))
        ).toMatchSnapshot();
      });
    });
  });
});
