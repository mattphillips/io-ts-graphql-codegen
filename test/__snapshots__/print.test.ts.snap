// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Print Input should inline named types name value 1`] = `
t.type({ a: A });

`;

exports[`Print Input should inline non-named types 1`] = `
t.type({ a: t.literal("A") });

`;

exports[`Print Mutation should inline named types name value 1`] = `
t.type({ a: A });

`;

exports[`Print Mutation should inline non-named types 1`] = `
t.type({ a: t.literal("A") });

`;

exports[`Print Named ObjectType should inline named types name value with __typename 1`] = `
t.type({
  __typename: t.literal("Alphabet"),
  a: A,
});

`;

exports[`Print Named ObjectType should inline non-named types with __typename 1`] = `
t.type({
  __typename: t.literal("Alphabet"),
  a: t.literal("A"),
});

`;

exports[`Print ObjectType should inline named types name value 1`] = `
t.type({ a: A });

`;

exports[`Print ObjectType should inline non-named types 1`] = `
t.type({ a: t.literal("A") });

`;

exports[`Print Query should inline named types name value 1`] = `
t.type({ a: A });

`;

exports[`Print Query should inline non-named types 1`] = `
t.type({ a: t.literal("A") });

`;

exports[`Print printNamedType InputType should remove option types with null union on array primitives 1`] = `
export const Name = t.type({ a: t.array(t.union([t.null, t.string])) });
export type Name = t.TypeOf<typeof Name>;

`;

exports[`Print printNamedType InputType should remove option types with null union on custom inputs 1`] = `
export const Name = t.type({ a: t.union([t.null, A]) });
export type Name = t.TypeOf<typeof Name>;

`;

exports[`Print printNamedType InputType should remove option types with null union on primitives 1`] = `
export const Name = t.type({ a: t.union([t.null, t.string]) });
export type Name = t.TypeOf<typeof Name>;

`;
